staging/vhost-device-can/src/vhu_can.rs

```diff
@@ -389,34 +403,44 @@ impl VhostUserCanBackend {
         }
 
         for desc_chain in requests {
-            let descriptors: Vec<_> = desc_chain.clone().collect();
-
-            if descriptors.len() != 2 {
-                return Err(Error::UnexpectedDescriptorCount(descriptors.len()));
-            }
-
-            trace!("descriptors.len(): {:?}", descriptors.len());
+            let atomic_mem = self.mem.as_ref().unwrap().memory();
 
-            let desc_request = descriptors[0];
-            if desc_request.is_write_only() {
-                warn!("Error::UnexpectedWriteOnlyDescriptor");
+            let Ok(mut reader) = desc_chain.clone().reader(&atomic_mem) else {
+                trace!("Error::UnexpectedWriteOnlyDescriptor");
                 return Err(Error::UnexpectedWriteOnlyDescriptor(0));
-            }
+            };
 
-            if desc_request.len() as usize != size_of::<VirtioCanCtrlRequest>() {
+            let reader_len = reader.available_bytes();
+            if reader_len < size_of::<VirtioCanCtrlRequest>() {
                 trace!(
                     "UnexpectedCtrlDescriptorSize, len = {:?}",
-                    desc_request.len()
+                    reader_len as u32
                 );
                 return Err(Error::UnexpectedCtrlDescriptorSize(
                     size_of::<VirtioCanCtrlRequest>(),
-                    desc_request.len(),
+                    reader_len as u32,
                 ));
             }
 
-            let request = desc_chain
-                .memory()
-                .read_obj::<VirtioCanCtrlRequest>(desc_request.addr())
+            let Ok(mut writer) = desc_chain.clone().writer(&atomic_mem) else {
+                trace!("Error::UnexpectedReadableDescriptor");
+                return Err(Error::UnexpectedReadableDescriptor(1));
```

Comment:
I think you can rewrite this with the correct error as I explained in a comment below.